huff0/_generate/gen.go

```diff
@@ -0,0 +1,347 @@
+	Pragma("noescape")
+
+	exhausted := GP64()
+	XORQ(exhausted.As64(), exhausted.As64()) // exhausted = false
+
+	limitPtr := AllocLocal(8)
+
+	bufferOrigin := GP64()
+	peekBits := GP64()
+	buffer := GP64()
+	dstEvery := GP64()
+	table := GP64()
+
+	br0 := GP64()
+	br1 := GP64()
+	br2 := GP64()
+	br3 := GP64()
+
+	Comment("Preload values")
+	{
+		ctx := Dereference(Param("ctx"))
+		Load(ctx.Field("peekBits"), peekBits)
+		Load(ctx.Field("out"), buffer)
+		MOVQ(buffer, bufferOrigin)
+		limit := Load(ctx.Field("limit"), GP64())
+		MOVQ(limit, limitPtr)
+		Load(ctx.Field("dstEvery"), dstEvery)
+		Load(ctx.Field("tbl"), table)
+		Load(ctx.Field("pbr0"), br0)
+		Load(ctx.Field("pbr1"), br1)
+		Load(ctx.Field("pbr2"), br2)
+		Load(ctx.Field("pbr3"), br3)
+	}
+
+	Comment("Main loop")
+	Label("main_loop")
+
+	MOVQ(bufferOrigin, buffer)
+	// Check if we have space
+	CMPQ(buffer, limitPtr)
+	SETGE(exhausted.As8())
+	d.decodeTwoValues(0, br0, peekBits, table, buffer, exhausted)
+	ADDQ(dstEvery, buffer)
+	d.decodeTwoValues(1, br1, peekBits, table, buffer, exhausted)
+	ADDQ(dstEvery, buffer)
+	d.decodeTwoValues(2, br2, peekBits, table, buffer, exhausted)
+	ADDQ(dstEvery, buffer)
+	d.decodeTwoValues(3, br3, peekBits, table, buffer, exhausted)
+
+	ADDQ(U8(2), bufferOrigin) // off += 2
+
+	TESTB(exhausted.As8(), exhausted.As8()) // any br[i].ofs < 4?
+	JZ(LabelRef("main_loop"))
+
+	{
+		ctx := Dereference(Param("ctx"))
+		tmp := Load(ctx.Field("out"), GP64())
+		decoded := GP64()
+		MOVQ(bufferOrigin, decoded)
+		SUBQ(tmp, decoded)
+		SHLQ(U8(2), decoded) // decoded *= 4
+
+		Store(decoded, ctx.Field("decoded"))
+	}
+
+	RET()
+}
+
+// TODO [wmu]: I believe it's doable in avo, but can't figure out how to deal
+//             with arbitrary pointers to a given type
+const bitReader_in = 0
+const bitReader_off = bitReader_in + 3*8 // {ptr, len, cap}
+const bitReader_value = bitReader_off + 8
+const bitReader_bitsRead = bitReader_value + 8
+
+func (d decompress4x) decodeTwoValues(id int, br, peekBits, table, buffer, exhausted reg.GPVirtual) {
+	brValue, brBitsRead := d.fillFast32(id, 32, br, exhausted)
+
+	val := GP64()
+	Commentf("val0 := br%d.peekTopBits(peekBits)", id)
+	CX := reg.CL
+	if d.bmi2 {
+		SHRXQ(peekBits, brValue, val.As64()) // val = (value >> peek_bits) & mask
+	} else {
+		MOVQ(brValue, val.As64())
+		MOVQ(peekBits, CX.As64())
+		SHRQ(CX, val.As64()) // val = (value >> peek_bits) & mask
+	}
+
+	Comment("v0 := table[val0&mask]")
+	v := reg.RDX
+	MOVW(Mem{Base: table, Index: val.As64(), Scale: 2}, v.As16())
+
+	Commentf("br%d.advance(uint8(v0.entry)", id)
+	out := reg.RAX            // Fixed since we need 8H
+	MOVB(v.As8H(), out.As8()) // BL = uint8(v0.entry >> 8)
+
+	MOVBQZX(v.As8(), CX.As64())
+	if d.bmi2 {
+		SHLXQ(v.As64(), brValue, brValue) // value <<= n
+	} else {
+		SHLQ(CX, brValue) // value <<= n
+	}
+
+	ADDQ(CX.As64(), brBitsRead) // bits_read += n
```

Comment:
```suggestion
	MOVW(Mem{Base: table, Index: val.As64(), Scale: 2}, CX.As16())

	Commentf("br%d.advance(uint8(v0.entry)", id)
	out := reg.RAX             // Fixed since we need 8H
	MOVB(CX.As8H(), out.As8()) // BL = uint8(v0.entry >> 8)

	SHLQ(CX, brValue)                // value <<= n
	ADDB(CX.As8(), brBitsRead.As8()) // bits_read += n
```